.cirrus.yml

```diff
@@ -115,7 +115,9 @@ task:
 
 task:
   name: 'ARM, unit tests, no functional tests'
+  # Skip if no ARM worker exists (for a fork)
   << : *GLOBAL_TASK_TEMPLATE
+  skip: $NO_ARM == "true"
```

Comment:
Is that new behavior? Let me test that again...

Reply:
I guess I never tested this, because the documentation just says:

```
# - arm64: For an aarch64 machine, recommended to have 2 CPUs and 8 GB of memory.
```

If the performance difference isn't too terrible, then I can just drop the commit and mention this in the documentation.

Reply:
`qemu-aarch64` adds quite a few packages to Ubuntu, so I'd probably want to keep the `NO_ARM` option around, but document this alternative.

Reply:
I guess this involves partially reverting #28087 since it seems that stripped the `qemu-arm` call from `00_setup_env_arm.sh`.

Reply:
(comments came in out of order for me)

I don't understand Docker, Podman and/or Qemu well enough to figure this out myself. For now I'll just leave this commit.

Reply:
I'm fine with just skipping that action on the stratum v2 based fork.

It's probably not too much hassle if there's clear instructions on how to do that on e.g. an AMD based Ubuntu machine.

The other way around doesn't work for me, because I don't have a native ARM machine (mac or otherwise). I'll eventually have one, a laptop. I might then use that for just the ARM task, not the rest of CI.

> moving it to GHA

I don't know how common ARM is for servers these days? I'm assuming it's mostly used by Apple, and for very low performance devices like Raspberry Pi's.

Reply:
After `podman run` is there some VM running that I have to login to, install the Cirrus job runner? Or is there some change needed to `00_setup_env_arm.sh` to make it use this?

Reply:
Thanks. I currently use `podman-docker` on Ubuntu 24.04. I created several `cirrus{1,2,3,4,5}` users.

For `cirrus1` I set `type: arm64`. I turned off `NO_ARM=true` for the occasion.

I installed `qemu-user-static` (`bash` and `podman-docker` were already there).

I ran your `podman run ...` incantation as the `cirrus1` user. And then started the worker:

```
RESTART_CI_DOCKER_BEFORE_RUN=1  USER_MODE=true MAKEJOBS=-j10 cirrus worker run -f ~/.cirrus.yaml
```

I then force-pushed an update to https://github.com/Sjors/bitcoin/pull/30. The ARM task fails: https://cirrus-ci.com/task/6683160417665024

```
STEP 6/7: RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]
exec container process `/usr/bin/bash`: Exec format error
Error: building at STEP "RUN bash -c cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh": while running runtime: exit status 1
```

Reply:
The CPU is an AMD Ryzen 9 7950X.

Reply:
That first command:

```
WARNING: image platform (linux/arm64/v8) does not match the expected platform (linux/amd64)
{"msg":"exec container process `/usr/bin/uname`: Exec format error","level":"error","time":"2024-06-20T09:03:31.220776Z"}
```

Running the second command does not change anything.

Reply:
```
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7950X 16-Core Processor
Virtualization:                       AMD-V
VERSION="24.04 LTS (Noble Numbat)"
podman version 4.9.3
```

Reply:
```
podman run --rm --privileged docker.io/multiarch/qemu-user-static --reset -p yes
Setting /usr/bin/qemu-alpha-static as binfmt interpreter for alpha
Setting /usr/bin/qemu-arm-static as binfmt interpreter for arm
Setting /usr/bin/qemu-armeb-static as binfmt interpreter for armeb
Setting /usr/bin/qemu-sparc-static as binfmt interpreter for sparc
Setting /usr/bin/qemu-sparc32plus-static as binfmt interpreter for sparc32plus
Setting /usr/bin/qemu-sparc64-static as binfmt interpreter for sparc64
Setting /usr/bin/qemu-ppc-static as binfmt interpreter for ppc
Setting /usr/bin/qemu-ppc64-static as binfmt interpreter for ppc64
Setting /usr/bin/qemu-ppc64le-static as binfmt interpreter for ppc64le
Setting /usr/bin/qemu-m68k-static as binfmt interpreter for m68k
Setting /usr/bin/qemu-mips-static as binfmt interpreter for mips
Setting /usr/bin/qemu-mipsel-static as binfmt interpreter for mipsel
Setting /usr/bin/qemu-mipsn32-static as binfmt interpreter for mipsn32
Setting /usr/bin/qemu-mipsn32el-static as binfmt interpreter for mipsn32el
Setting /usr/bin/qemu-mips64-static as binfmt interpreter for mips64
Setting /usr/bin/qemu-mips64el-static as binfmt interpreter for mips64el
Setting /usr/bin/qemu-sh4-static as binfmt interpreter for sh4
Setting /usr/bin/qemu-sh4eb-static as binfmt interpreter for sh4eb
Setting /usr/bin/qemu-s390x-static as binfmt interpreter for s390x
Setting /usr/bin/qemu-aarch64-static as binfmt interpreter for aarch64
Setting /usr/bin/qemu-aarch64_be-static as binfmt interpreter for aarch64_be
Setting /usr/bin/qemu-hppa-static as binfmt interpreter for hppa
Setting /usr/bin/qemu-riscv32-static as binfmt interpreter for riscv32
Setting /usr/bin/qemu-riscv64-static as binfmt interpreter for riscv64
Setting /usr/bin/qemu-xtensa-static as binfmt interpreter for xtensa
Setting /usr/bin/qemu-xtensaeb-static as binfmt interpreter for xtensaeb
Setting /usr/bin/qemu-microblaze-static as binfmt interpreter for microblaze
Setting /usr/bin/qemu-microblazeel-static as binfmt interpreter for microblazeel
Setting /usr/bin/qemu-or1k-static as binfmt interpreter for or1k
Setting /usr/bin/qemu-hexagon-static as binfmt interpreter for hexagon
```

Reply:
When I run both commands as root, it does work. But doing so does not change anything for the non-root user. At least that's a clue.

Reply:
Some background information on what this actually does (though not podman specific): https://stackoverflow.com/a/72890225

Reply:
Adding `cirrus1` to `/etc/sudoers` and running both commands with `sudo` works. But running only the `--reset` command with `sudo` is not enough.

Not really a solution, but at least I can run the CI job once to see what the performance is like.

Reply:
Uhhh ... so I did a reboot. And now `podman run --rm -ti "docker.io/arm64v8/debian:bookworm" uname --machine` just works, doesn't need the reset command.

Reply:
If the job finishes in reasonable time, I'll add something like this to the documentation: https://github.com/Sjors/bitcoin/commit/d065dc11fb862f48fc88e674a95b7daf68269ca0

It seems that this either works trivially or is really hard to debug, so I prefer to keep `NO_ARM=1` around as an alternative.

Reply:
1 hour 42 minutes, but that was without using any cache.

Reply:
58 minutes for the second run, which is acceptable.

Reply:
Without this change, when there's no ARM machine, it remains pending forever.

Reply:
> I understand. However, my point is that this is a cosmetic thing. Who cares? The person that sets up a fork, enables cirrus on it, enables a worker pool on it, knowingly does not enable an arm worker on it, so the person should know that the arm task will not run. Whatever the icon in the UI displays should thus not matter.

I agree with Marco. We seem to be spending a lot of time, and happy to take on some complexity just to accommodate use-cases *outside* this repository; for which, as far as I can see, there is not actually that much demand. We also shouldn't have to introduce complexity here, just to fix cosmetic issues in other repositories.

To me it's also unclear what the requirements are here going forward. I assume if any of the outside-of-this-repo functionality breaks, eventually someone might notice, and a fix might happen, but at the same time, I don't think anyone in this repo is going to be thinking about this when making code changes here (and they shouldn't have to), so things may break, and that will be unfortunate, but should never be blocking changes here. We are maintaining our infrastructure; not our infrastructure plus the infrastructure of whoever happened to fork this repository.

Reply:
It's confusing for both the pull request author and reviewers. Only the fork maintainer will know what's going on, but they also need to carefully check each time that it's only the expected job that's stuck. A single green check, when all jobs pass, is more clear.

> does yaml allow duplicate keys

Not sure.

So the simplest solution for now is to drop 859e1c558fd9c8604fda361e465aa4eb4a676823 and make qemu "mandatory", which works for me.

> just to accommodate use-cases outside this repository

But it reduces the number of direct pull requests _to_ repository, so on balance it may save time.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1652884524

I don't have a strong opinion, but I do think the NO_ARM commit 859e1c558fd9c8604fda361e465aa4eb4a676823 is a nice change because I don't think it fixes a purely cosmetic issue. I think it removes confusing behavior of opening a PR and seeing a job hang forever. The actual code change is very simple, just adding `skip: $NO_ARM == "true"` and I think the documentation change makes the whole "A self-hosted machine(s) can be used via Cirrus CI..." section clearer and more concrete. But I agree with dropping it as long as any objections remain.

Reply:
It's conceptually simple, but I agree that we shouldn't use duplicate keys unless Cirrus docs explicitly says you can. I could work around it in various ways, but for now I found it easier to just drop the commit.